src/Language/Haskell/TH/Syntax/Compat.hs

```diff
@@ -225,11 +272,11 @@ liftQuote = unsafeQToQuote . Syntax.lift
 -- This is a variant of the 'Syntax.liftTyped' function that is always
 -- guaranteed to use a 'Quote' constraint, even on old versions of
 -- @template-haskell@.
-liftTypedQuote :: (Syntax.Lift t, Quote m) => t -> m (Syntax.TExp t)
+liftTypedQuote :: (Syntax.Lift t, Quote m) => t -> Code m t
 # if MIN_VERSION_template_haskell(2,17,0)
 liftTypedQuote = Syntax.liftTyped
 # else
-liftTypedQuote = unsafeQToQuote . Syntax.liftTyped
+liftTypedQuote = liftCode . unsafeQToQuote . Syntax.liftTyped
```

Comment:
`liftTyped` is only since template-haskell-2.15?

Reply:
Do you mean `template-haskell-2.16.0.0`? (Notice the `#if MIN_VERSION_template_haskell(2,16,0)` part above.)

Reply:
```haskell
liftTypedQuote :: (TH.Lift a, Compat.Quote q) => a -> Code q a
#if MIN_VERSION_template_haskell(2,17,0)
liftTypedQuote = TH.liftTyped
#elif MIN_VERSION_template_haskell(2,16,0)
liftTypedQuote = liftCode . Compat.unsafeQToQuote  . Compat.lifTyped
#else
liftTypedQuote = unsafeCodeCoerce . Compat.liftQuote
#endif
```

is "correct" implementation. My test cases pass on GHC-8.8.

Reply:
Ah, you wish to go even further and backport `liftTypedQuote` back to GHC 7.0? That is certainly something that we could consider, but again, there's a somewhat tricky issue of conditional APIs here. Strictly speaking, it's possible that someone could define `lift` and `liftTyped` in completely different ways, and if they did so, then `liftTypedQuote` would produce different results depending on whether you use `template-haskell-2.16`+ or not.

Still, I can see the value of having this, much like I can see the value of having `Splice{Q}`. I just wonder if we should attach this conditional-API behavior to `liftTypedQuote`, or a separate function altogether.

Reply:
> Strictly speaking, it's possible that someone could define `lift` and `liftTyped` in completely different ways.

These instances should burn. If there isn't a law `lift = unTypeCode/Q . liftTyped`, we should add one.

I consider it as defining `Eq` with incompatible `==` and `/=`.

Reply:
Ah, good point. Now that I've went and checked, [the Haddocks for `Lift`](https://hackage.haskell.org/package/template-haskell-2.16.0.0/docs/Language-Haskell-TH-Syntax.html#t:Lift) _do_ require `lift x ≡ unTypeQ (liftTyped x)`. In light of this, I've decided to use your version of `liftTypedQuote` from https://github.com/haskell-compat/th-quote/pull/1#discussion_r463956577, and I've added a brief disclaimer in its Haddocks about what might happen if you use an unlawful `Lift` instance.

One minor correction to my comment in https://github.com/haskell-compat/th-quote/pull/1#discussion_r463970433: your version of `liftTypedQuote` can only be backported back to GHC 7.8 (not 7.0), since that's when typed TH was first introduced. (I suppose we _could_ backport `Code`/`TExp` back to older versions of GHC, but there wouldn't be much point in doing so.)